Collection literals working group meeting 2023-05-26 (#7232)
dotnet/csharplang: meetings/working-groups/collection-literals/CL-2023-05-26.md

This would work fine when the constructed collection has a writeable Capacity property because codegen could set the capacity before calling `Add(T)` in a loop. However, if there's no writeable Capacity property, we're better off calling `AddRange` and boxing, because the `AddRange` implementation is the only way to intelligently avoid intermediate resizes. (In this situation, the thing being spread has a known length but the collection literal as a whole has an unknown length due to an earlier spread.)

#### Conclusion (unknown-length construction)

Stick with building using Add and AddRange. This is good enough for 95% of customers. Allow the compiler to special-case codegen for widely-used collections to improve performance.

The compiler should prefer calling `AddRange(ReadOnlySpan<T>)` if such a method exists and the thing being spread is a span or exposes a span.

### Known-length construction

> The ideal API for constructing collections with contiguous backing storage is to pass the capacity and receive a span to the backing storage. To enable creation methods to be in the same class while having the same name, the collection could be passed back using an `out` parameter.

Reviewer: >To enable creation methods to be in the same class while having the same name [](http://example.com/codeflow?start=142&length=77)

I initially interpreted "to be in the same class" as "in the same class _as the collection being created_".

I'm not sure how to make it clearer, but perhaps: "To allow multiple creation methods to be declared in the same class with the same name, ..."

> The type parameter to use when calling the creation method could be taken from the generic type argument of the collection's `IEnumerable<>` implementation. This strategy would not work for collections which implement `IEnumerable<>` more than once.

Author: What do y'all think about using the concept of the enumerated type of the collection type instead of specifically talking about IEnumerable implementations? I.e. the type that would become the type of the iteration variable in a foreach over the collection type being constructed?

Reviewer 2: That definitely fits the intuition/wording elsewhere. We've called that the 'iteration type' of a particular value.

Author: ```suggestion
The type parameter to use when calling the creation method could be the iteration type of the collection being constructed.
```

```cs
[CollectionLiteralBuilder(
    typeof(CollectionsMarshal),
    nameof(CollectionsMarshal.CreateUnsafe))]
public class MyCollection<T> : IEnumerable<T> { }

public static class CollectionsMarshal
{
    public static void CreateUnsafe<T>(
        int capacity,
        out MyCollection<T> collection,
        out Span<T> storage);

    public static void CreateUnsafe<T>(
        int capacity,
        out List<T> collection,
        out Span<T> storage);
}
```

The reverse API pattern (passing a span to the collection rather than receiving one) would enable the construction of collections that have noncontiguous backing storage and do not have mutating Add/AddRange methods, such as ImmutableDictionary.

This would allow the use of existing Create methods in the BCL:

```cs
[CollectionLiteralBuilder(
    typeof(MyCollection),
    nameof(MyCollection.Create))]
public class MyCollection<T> : IEnumerable<T> { }

public static class MyCollection
{
    public static MyCollection<T> Create<T>(ReadOnlySpan<T> elements);
}
```

#### Conclusion (known-length construction)

> Use an attribute to point to a creation method similar to the code samples above. If this attribute is present, prefer it over generating Add/AddRange calls.

Author: I need different headings, because this also applies to unknown-length collection literals.
